I can't seem to figure out why this is failing to build with this error:

`.../WebKitBuild/Debug/DerivedSources/WebKit/NetworkProcessMessages.h:52:10: fatal error:`
`'enum:bool WebKit/LocalStorageType.h' file not found`
   `52 | #include <enum:bool WebKit/LocalStorageType.h>`
`1 error generated.`

--- a/Source/WebKit/NetworkProcess/NetworkProcess.cpp
+++ b/Source/WebKit/NetworkProcess/NetworkProcess.cpp
@@ -3116,4 +3116,15 @@ void NetworkProcess::fetchLocalStorage(PAL::SessionID sessionID, CompletionHandl
     session->protectedStorageManager()->fetchLocalStorage(WTFMove(completionHandler));
 }
 
+void NetworkProcess::restoreLocalStorage(PAL::SessionID sessionID, HashMap<WebCore::ClientOrigin, HashMap<String, String>>&& localStorageMap, CompletionHandler<void(bool)>&& completionHandler)
+{
+    CheckedPtr session = networkSession(sessionID);
+    if (!session) {
+        completionHandler(false);
+        return;
+    }
+
+    session->protectedStorageManager()->restoreLocalStorage(WTFMove(localStorageMap), WTFMove(completionHandler));
+}
+
 } // namespace WebKit

--- a/Source/WebKit/NetworkProcess/NetworkProcess.h
+++ b/Source/WebKit/NetworkProcess/NetworkProcess.h
@@ -453,6 +453,7 @@ public:
     bool enableModernDownloadProgress() const { return m_enableModernDownloadProgress; }
 
     void fetchLocalStorage(PAL::SessionID, CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&)>&&);
+    void restoreLocalStorage(PAL::SessionID, HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&, CompletionHandler<void(bool)>&&);
 
 private:
     void platformInitializeNetworkProcess(const NetworkProcessCreationParameters&);

--- a/Source/WebKit/NetworkProcess/NetworkProcess.messages.in
+++ b/Source/WebKit/NetworkProcess/NetworkProcess.messages.in
@@ -261,4 +261,5 @@ messages -> NetworkProcess : AuxiliaryProcess WantsAsyncDispatchMessage {
     GetAppBadgeForTesting(PAL::SessionID sessionID) -> (std::optional<uint64_t> badge)
 
     FetchLocalStorage(PAL::SessionID sessionID) -> (HashMap<WebCore::ClientOrigin, HashMap<String, String>> localStorageMap)
+    RestoreLocalStorage(PAL::SessionID sessionID, HashMap<WebCore::ClientOrigin, HashMap<String, String>> localStorageMap) -> (bool succeeded)
 }

--- a/Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp
+++ b/Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp
@@ -185,7 +185,7 @@ void LocalStorageManager::connectionClosedForTransientStorageArea(IPC::Connectio
     m_transientStorageArea = nullptr;
 }
 
-StorageAreaIdentifier LocalStorageManager::connectToLocalStorageArea(IPC::Connection::UniqueID connection, StorageAreaMapIdentifier sourceIdentifier, const WebCore::ClientOrigin& origin, Ref<WorkQueue>&& workQueue)
+StorageAreaBase& LocalStorageManager::ensureLocalStorageArea(const WebCore::ClientOrigin& origin, Ref<WorkQueue>&& workQueue)
 {
     if (!m_localStorageArea) {
         RefPtr<StorageAreaBase> storage;
@@ -198,17 +198,18 @@ StorageAreaIdentifier LocalStorageManager::connectToLocalStorageArea(IPC::Connec
         m_registry->registerStorageArea(storage->identifier(), *storage);
     }
 
-    ASSERT(m_path.isEmpty() || is<SQLiteStorageArea>(*m_localStorageArea));
-    protectedLocalStorageArea()->addListener(connection, sourceIdentifier);
-    return m_localStorageArea->identifier();
+    return *m_localStorageArea;
 }
 
-RefPtr<StorageAreaBase> LocalStorageManager::protectedLocalStorageArea() const
+StorageAreaIdentifier LocalStorageManager::connectToLocalStorageArea(IPC::Connection::UniqueID connection, StorageAreaMapIdentifier sourceIdentifier, const WebCore::ClientOrigin& origin, Ref<WorkQueue>&& workQueue)
 {
-    return m_localStorageArea;
+    Ref localStorageArea = ensureLocalStorageArea(origin, WTFMove(workQueue));
+    ASSERT(m_path.isEmpty() || is<SQLiteStorageArea>(localStorageArea));
+    localStorageArea->addListener(connection, sourceIdentifier);
+    return localStorageArea->identifier();
 }
 
-StorageAreaIdentifier LocalStorageManager::connectToTransientLocalStorageArea(IPC::Connection::UniqueID connection, StorageAreaMapIdentifier sourceIdentifier, const WebCore::ClientOrigin& origin)
+MemoryStorageArea& LocalStorageManager::ensureTransientLocalStorageArea(const WebCore::ClientOrigin& origin)
 {
     RefPtr transientStorageArea = m_transientStorageArea;
     if (!transientStorageArea) {
@@ -217,7 +218,13 @@ StorageAreaIdentifier LocalStorageManager::connectToTransientLocalStorageArea(IP
         m_registry->registerStorageArea(transientStorageArea->identifier(), *transientStorageArea);
     }
 
-    ASSERT(is<MemoryStorageArea>(*transientStorageArea));
+    return *m_transientStorageArea;
+}
+
+StorageAreaIdentifier LocalStorageManager::connectToTransientLocalStorageArea(IPC::Connection::UniqueID connection, StorageAreaMapIdentifier sourceIdentifier, const WebCore::ClientOrigin& origin)
+{
+    Ref transientStorageArea = ensureTransientLocalStorageArea(origin);
+    ASSERT(is<MemoryStorageArea>(transientStorageArea));
     transientStorageArea->addListener(connection, sourceIdentifier);
     return transientStorageArea->identifier();
 }
@@ -252,6 +259,27 @@ HashMap<String, String> LocalStorageManager::fetchStorageMap() const
         return transientStorageArea->allItems();
 
     return { };
+}
+
+bool LocalStorageManager::populateStorageArea(WebCore::ClientOrigin clientOrigin, HashMap<String, String>&& storageMap, Ref<WorkQueue>&& workQueue)
+{
+    bool succeeded = true;
+
+    if (clientOrigin.topOrigin == clientOrigin.clientOrigin) {
+        Ref localStorageArea = ensureLocalStorageArea(clientOrigin, WTFMove(workQueue));
+        for (auto& [key, value] : storageMap) {
+            if (!localStorageArea->setItem({ }, { }, WTFMove(key), WTFMove(value), { }))
+                succeeded = false;
+        }
+    } else {
+        Ref transientStorageArea = ensureTransientLocalStorageArea(clientOrigin);
+        for (auto& [key, value] : storageMap) {
+            if (!transientStorageArea->setItem({ }, { }, WTFMove(key), WTFMove(value), { }))
+                succeeded = false;
+        }
+    }
+
+    return succeeded;
 }
 
 } // namespace WebKit

--- a/Source/WebKit/NetworkProcess/storage/LocalStorageManager.h
+++ b/Source/WebKit/NetworkProcess/storage/LocalStorageManager.h
@@ -66,12 +66,14 @@ public:
     void disconnectFromStorageArea(IPC::Connection::UniqueID, StorageAreaIdentifier);
 
     HashMap<String, String> fetchStorageMap() const;
+    bool populateStorageArea(WebCore::ClientOrigin, HashMap<String, String>&&, Ref<WorkQueue>&&);
 
 private:
     void connectionClosedForLocalStorageArea(IPC::Connection::UniqueID);
     void connectionClosedForTransientStorageArea(IPC::Connection::UniqueID);
 
-    RefPtr<StorageAreaBase> protectedLocalStorageArea() const;
+    StorageAreaBase& ensureLocalStorageArea(const WebCore::ClientOrigin&, Ref<WorkQueue>&&);
+    MemoryStorageArea& ensureTransientLocalStorageArea(const WebCore::ClientOrigin&);
 
     String m_path;
     CheckedRef<StorageAreaRegistry> m_registry;

--- a/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.cpp
+++ b/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.cpp
@@ -59,7 +59,7 @@ HashMap<String, String> MemoryStorageArea::allItems()
     return m_map.items();
 }
 
-Expected<void, StorageError> MemoryStorageArea::setItem(IPC::Connection::UniqueID connection, StorageAreaImplIdentifier storageAreaImplID, String&& key, String&& value, const String& urlString)
+Expected<void, StorageError> MemoryStorageArea::setItem(std::optional<IPC::Connection::UniqueID> connection, std::optional<StorageAreaImplIdentifier> storageAreaImplID, String&& key, String&& value, const String& urlString)
 {
     String oldValue;
     bool hasQuotaError = false;
@@ -67,7 +67,8 @@ Expected<void, StorageError> MemoryStorageArea::setItem(IPC::Connection::UniqueI
     if (hasQuotaError)
         return makeUnexpected(StorageError::QuotaExceeded);
 
-    dispatchEvents(connection, storageAreaImplID, key, oldValue, value, urlString);
+    if (connection && storageAreaImplID)
+        dispatchEvents(*connection, *storageAreaImplID, key, oldValue, value, urlString);
 
     return { };
 }

--- a/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h
+++ b/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h
@@ -47,6 +47,7 @@ public:
 
     // StorageAreaBase
     HashMap<String, String> allItems() final;
+    Expected<void, StorageError> setItem(std::optional<IPC::Connection::UniqueID>, std::optional<StorageAreaImplIdentifier>, String&& key, String&& value, const String& urlString) final;
     
     void ref() const final { RefCounted::ref(); }
     void deref() const final { RefCounted::deref(); }
@@ -56,7 +57,6 @@ private:
 
     // StorageAreaBase
     StorageAreaBase::StorageType storageType() const final { return m_storageType; }
-    Expected<void, StorageError> setItem(IPC::Connection::UniqueID, StorageAreaImplIdentifier, String&& key, String&& value, const String& urlString) final;
     Expected<void, StorageError> removeItem(IPC::Connection::UniqueID, StorageAreaImplIdentifier, const String& key, const String& urlString) final;
     Expected<void, StorageError> clear(IPC::Connection::UniqueID, StorageAreaImplIdentifier, const String& urlString) final;
 

--- a/Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp
+++ b/Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp
@@ -1332,11 +1332,35 @@ void NetworkStorageManager::fetchLocalStorage(CompletionHandler<void(HashMap<Web
             auto storageMap = localStorageManager.fetchStorageMap();
 
             if (!storageMap.isEmpty())
-                localStorageMap.add(origin, storageMap);
+                localStorageMap.add(origin, WTFMove(storageMap));
         }
 
-        RunLoop::protectedMain()->dispatch([completionHandler = WTFMove(completionHandler), localStorageMap = crossThreadCopy(WTFMove(localStorageMap))] mutable {
+        RunLoop::protectedMain()->dispatch([protectedThis = WTFMove(protectedThis), completionHandler = WTFMove(completionHandler), localStorageMap = crossThreadCopy(WTFMove(localStorageMap))] mutable {
             completionHandler(WTFMove(localStorageMap));
+        });
+    });
+}
+
+void NetworkStorageManager::restoreLocalStorage(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&& localStorageMap, CompletionHandler<void(bool)>&& completionHandler)
+{
+    ASSERT(RunLoop::isMain());
+    ASSERT(!m_closed);
+
+    protectedWorkQueue()->dispatch([this, protectedThis = Ref { *this }, localStorageMap = crossThreadCopy(WTFMove(localStorageMap)), completionHandler = WTFMove(completionHandler)]() mutable {
+        assertIsCurrent(workQueue());
+
+        bool succeeded = true;
+
+        for (auto& [clientOrigin, storageMap] : localStorageMap) {
+            auto& localStorageManager = originStorageManager(clientOrigin, ShouldWriteOriginFile::Yes).localStorageManager(*m_storageAreaRegistry);
+            auto result = localStorageManager.populateStorageArea(clientOrigin, WTFMove(storageMap), protectedWorkQueue());
+
+            if (!result)
+                succeeded = false;
+        }
+
+        RunLoop::protectedMain()->dispatch([protectedThis = WTFMove(protectedThis), completionHandler = WTFMove(completionHandler), succeeded] mutable {
+            completionHandler(succeeded);
         });
     });
 }

--- a/Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h
+++ b/Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h
@@ -125,6 +125,7 @@ public:
     void handleLowMemoryWarning();
     void syncLocalStorage(CompletionHandler<void()>&&);
     void fetchLocalStorage(CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&)>&&);
+    void restoreLocalStorage(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&, CompletionHandler<void(bool)>&&);
     void registerTemporaryBlobFilePaths(IPC::Connection&, const Vector<String>&);
     void requestSpace(const WebCore::ClientOrigin&, uint64_t size, CompletionHandler<void(bool)>&&);
     void resetQuotaForTesting(CompletionHandler<void()>&&);

--- a/Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp
+++ b/Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp
@@ -341,7 +341,7 @@ HashMap<String, String> SQLiteStorageArea::allItems()
     return items;
 }
 
-Expected<void, StorageError> SQLiteStorageArea::setItem(IPC::Connection::UniqueID connection, StorageAreaImplIdentifier storageAreaImplID, String&& key, String&& value, const String& urlString)
+Expected<void, StorageError> SQLiteStorageArea::setItem(std::optional<IPC::Connection::UniqueID> connection, std::optional<StorageAreaImplIdentifier> storageAreaImplID, String&& key, String&& value, const String& urlString)
 {
     ASSERT(!isMainRunLoop());
 
@@ -379,7 +379,8 @@ Expected<void, StorageError> SQLiteStorageArea::setItem(IPC::Connection::UniqueI
         return makeUnexpected(StorageError::Database);
     }
 
-    dispatchEvents(connection, storageAreaImplID, key, oldValue, value, urlString);
+    if (connection && storageAreaImplID)
+        dispatchEvents(*connection, *storageAreaImplID, key, oldValue, value, urlString);
     updateCacheIfNeeded(key, value);
 
     return { };

--- a/Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.h
+++ b/Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.h
@@ -59,7 +59,7 @@ private:
     StorageType storageType() const final { return StorageAreaBase::StorageType::Local; };
     bool isEmpty() final;
     HashMap<String, String> allItems() final;
-    Expected<void, StorageError> setItem(IPC::Connection::UniqueID, StorageAreaImplIdentifier, String&& key, String&& value, const String& urlString) final;
+    Expected<void, StorageError> setItem(std::optional<IPC::Connection::UniqueID>, std::optional<StorageAreaImplIdentifier>, String&& key, String&& value, const String& urlString) final;
     Expected<void, StorageError> removeItem(IPC::Connection::UniqueID, StorageAreaImplIdentifier, const String& key, const String& urlString) final;
     Expected<void, StorageError> clear(IPC::Connection::UniqueID, StorageAreaImplIdentifier, const String& urlString) final;
 

--- a/Source/WebKit/NetworkProcess/storage/StorageAreaBase.h
+++ b/Source/WebKit/NetworkProcess/storage/StorageAreaBase.h
@@ -69,7 +69,7 @@ public:
     void notifyListenersAboutClear();
 
     virtual HashMap<String, String> allItems() = 0;
-    virtual Expected<void, StorageError> setItem(IPC::Connection::UniqueID, StorageAreaImplIdentifier, String&& key, String&& value, const String& urlString) = 0;
+    virtual Expected<void, StorageError> setItem(std::optional<IPC::Connection::UniqueID>, std::optional<StorageAreaImplIdentifier>, String&& key, String&& value, const String& urlString) = 0;
     virtual Expected<void, StorageError> removeItem(IPC::Connection::UniqueID, StorageAreaImplIdentifier, const String& key, const String& urlString) = 0;
     virtual Expected<void, StorageError> clear(IPC::Connection::UniqueID, StorageAreaImplIdentifier, const String& urlString) = 0;
 

--- a/Source/WebKit/Shared/WebsiteData/WebsiteDataType.h
+++ b/Source/WebKit/Shared/WebsiteData/WebsiteDataType.h
@@ -53,3 +53,35 @@ enum class WebsiteDataType : uint32_t {
 };
 
 } // namespace WebKit
+
+namespace WTF {
+
+template<> struct EnumTraitsForPersistence<WebKit::WebsiteDataType> {
+    using values = EnumValues<
+        WebKit::WebsiteDataType,
+        WebKit::WebsiteDataType::Cookies,
+        WebKit::WebsiteDataType::DiskCache,
+        WebKit::WebsiteDataType::MemoryCache,
+        WebKit::WebsiteDataType::OfflineWebApplicationCache,
+        WebKit::WebsiteDataType::SessionStorage,
+        WebKit::WebsiteDataType::LocalStorage,
+        WebKit::WebsiteDataType::WebSQLDatabases,
+        WebKit::WebsiteDataType::IndexedDBDatabases,
+        WebKit::WebsiteDataType::MediaKeys,
+        WebKit::WebsiteDataType::HSTSCache,
+        WebKit::WebsiteDataType::SearchFieldRecentSearches,
+        WebKit::WebsiteDataType::ResourceLoadStatistics,
+        WebKit::WebsiteDataType::Credentials,
+        WebKit::WebsiteDataType::ServiceWorkerRegistrations,
+        WebKit::WebsiteDataType::DOMCache,
+        WebKit::WebsiteDataType::DeviceIdHashSalt,
+        WebKit::WebsiteDataType::PrivateClickMeasurements,
+#if HAVE(ALTERNATIVE_SERVICE)
+        WebKit::WebsiteDataType::AlternativeServices,
+#endif
+        WebKit::WebsiteDataType::FileSystem,
+        WebKit::WebsiteDataType::BackgroundFetchStorage
+    >;
+};
+
+} // namespace WTF

--- a/Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm
+++ b/Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm
@@ -63,6 +63,7 @@
 #import <WebCore/WebCoreObjCExtras.h>
 #import <WebCore/WebCorePersistentCoders.h>
 #import <wtf/BlockPtr.h>
+#import <wtf/CallbackAggregator.h>
 #import <wtf/TZoneMallocInlines.h>
 #import <wtf/URL.h>
 #import <wtf/Vector.h>
@@ -70,6 +71,8 @@
 #import <wtf/cocoa/RuntimeApplicationChecksCocoa.h>
 #import <wtf/cocoa/SpanCocoa.h>
 #import <wtf/cocoa/VectorCocoa.h>
+#import <wtf/persistence/PersistentDecoder.h>
+#import <wtf/persistence/PersistentEncoder.h>
 
 #if HAVE(NW_PROXY_CONFIG)
 #import <Network/Network.h>
@@ -1428,22 +1431,98 @@ static Vector<WebKit::WebsiteDataRecord> toWebsiteDataRecords(NSArray *dataRecor
     });
 }
 
-- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler
+struct WKWebsiteData {
+    std::optional<HashMap<WebCore::ClientOrigin, HashMap<String, String>>> localStorage;
+};
+
+static Vector<WebKit::WebsiteDataType> filterSupportedTypes(NSSet<NSString *> * dataTypes)
 {
+    Vector<WebKit::WebsiteDataType> supportedTypes;
+    if ([dataTypes containsObject:WKWebsiteDataTypeLocalStorage])
+        supportedTypes.append(WebKit::WebsiteDataType::LocalStorage);
+
+    return supportedTypes;
+}
+
+- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(void(^)(NSData *))completionHandler
+{
+    Vector<WebKit::WebsiteDataType> dataTypesToEncode = filterSupportedTypes(dataTypes);
+    auto data = Box<WKWebsiteData>::create();
+
+    auto callbackAggregator = CallbackAggregator::create([completionHandler = makeBlockPtr(completionHandler), dataTypesToEncode = WTFMove(dataTypesToEncode), data] {
+        WTF::Persistence::Encoder encoder;
+        constexpr unsigned currentWKWebsiteDataSerializationVersion = 1;
+        encoder << currentWKWebsiteDataSerializationVersion;
+        encoder << dataTypesToEncode;
+
+        for (auto& dataTypeToEncode : dataTypesToEncode) {
+            switch (dataTypeToEncode) {
+            case WebKit::WebsiteDataType::LocalStorage:
+                encoder << data->localStorage.value();
+                break;
+            default:
+                ASSERT_NOT_REACHED();
+                break;
+            }
+        }
+
+        completionHandler(toNSData(encoder.span()).get());
+    });
+
     if ([dataTypes containsObject:WKWebsiteDataTypeLocalStorage]) {
-        _websiteDataStore->fetchLocalStorage([completionHandler = makeBlockPtr(completionHandler)](HashMap<WebCore::ClientOrigin, HashMap<String, String>>&& localStorage) {
-            constexpr unsigned currentLocalStorageSerializationVersion = 1;
-
-            WTF::Persistence::Encoder encoder;
-            encoder << currentLocalStorageSerializationVersion;
-            encoder << localStorage;
-            completionHandler(toNSData(encoder.span()).get());
+        _websiteDataStore->fetchLocalStorage([callbackAggregator, data](HashMap<WebCore::ClientOrigin, HashMap<String, String>>&& localStorage) {
+            data->localStorage = WTFMove(localStorage);
         });
+    }
+}
 
+- (void)_restoreData:(NSData *)data completionHandler:(void(^)(BOOL))completionHandler
+{
+    WTF::Persistence::Decoder decoder(span(data));
+
+    std::optional<unsigned> currentLocalStorageSerializationVersion;
+    decoder >> currentLocalStorageSerializationVersion;
+    if (!currentLocalStorageSerializationVersion) {
+        completionHandler(NO);
         return;
     }
 
-    completionHandler(nil);
+    std::optional<Vector<WebKit::WebsiteDataType>> encodedDataTypes;
+    decoder >> encodedDataTypes;
+    if (!encodedDataTypes) {
+        completionHandler(NO);
+        return;
+    }
+
+    auto succeeded = Box<bool>::create(true);
+    auto callbackAggregator = CallbackAggregator::create([completionHandler = makeBlockPtr(completionHandler), succeeded] {
+        completionHandler(*succeeded);
+    });
+
+    for (auto& encodedDataType : *encodedDataTypes) {
+        switch (encodedDataType) {
+        case WebKit::WebsiteDataType::LocalStorage: {
+            std::optional<HashMap<WebCore::ClientOrigin, HashMap<String, String>>> localStorage;
+            decoder >> localStorage;
+
+            if (!localStorage) {
+                *succeeded = false;
+                return;
+            }
+
+            if (!localStorage->isEmpty()) {
+                _websiteDataStore->restoreLocalStorage(WTFMove(*localStorage), [callbackAggregator, succeeded](bool restoreSucceeded) {
+                    if (!restoreSucceeded)
+                        *succeeded = false;
+                });
+            }
+            break;
+        }
+        default:
+            ASSERT_NOT_REACHED();
+            break;
+        }
+    }
 }
 
 @end

--- a/Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h
+++ b/Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h
@@ -155,7 +155,9 @@ typedef NS_ENUM(uint8_t, _WKRestrictedOpenerType) {
 
 - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
 
-- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler;
+// Supported data types are: WKWebsiteDataTypeLocalStorage.
+- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void(^)(NSData *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
+- (void)_restoreData:(NSData *)data completionHandler:(WK_SWIFT_UI_ACTOR void(^)(BOOL))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
 
 @end
 

--- a/Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp
+++ b/Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp
@@ -2003,11 +2003,12 @@ void NetworkProcessProxy::setEmulatedConditions(PAL::SessionID sessionID, std::o
 
 void NetworkProcessProxy::fetchLocalStorage(PAL::SessionID sessionID, CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&)>&& completionHandler)
 {
-    if (!canSendMessage()) {
-        completionHandler({ });
-    }
-
     sendWithAsyncReply(Messages::NetworkProcess::FetchLocalStorage(sessionID), WTFMove(completionHandler));
+}
+
+void NetworkProcessProxy::restoreLocalStorage(PAL::SessionID sessionID, HashMap<WebCore::ClientOrigin, HashMap<String, String>>&& localStorage, CompletionHandler<void(bool)>&& completionHandler)
+{
+    sendWithAsyncReply(Messages::NetworkProcess::RestoreLocalStorage(sessionID, WTFMove(localStorage)), WTFMove(completionHandler));
 }
 
 } // namespace WebKit

--- a/Source/WebKit/UIProcess/Network/NetworkProcessProxy.h
+++ b/Source/WebKit/UIProcess/Network/NetworkProcessProxy.h
@@ -342,6 +342,7 @@ public:
     void notifyMediaStreamingActivity(bool);
 
     void fetchLocalStorage(PAL::SessionID, CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&)>&&);
+    void restoreLocalStorage(PAL::SessionID, HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&, CompletionHandler<void(bool)>&&);
 
 private:
     explicit NetworkProcessProxy();

--- a/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp
+++ b/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp
@@ -2716,9 +2716,14 @@ void WebsiteDataStore::setRestrictedOpenerTypeForDomainForTesting(const WebCore:
     m_restrictedOpenerTypesForTesting.set(domain, type);
 }
 
-void WebsiteDataStore::fetchLocalStorage(CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>)>&& completionHandler)
+void WebsiteDataStore::fetchLocalStorage(CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&)>&& completionHandler)
 {
     protectedNetworkProcess()->fetchLocalStorage(m_sessionID, WTFMove(completionHandler));
+}
+
+void WebsiteDataStore::restoreLocalStorage(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&& localStorage, CompletionHandler<void(bool)>&& completionHandler)
+{
+    protectedNetworkProcess()->restoreLocalStorage(m_sessionID, WTFMove(localStorage), WTFMove(completionHandler));
 }
 
 } // namespace WebKit

--- a/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h
+++ b/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h
@@ -483,7 +483,8 @@ public:
 
     void getAppBadgeForTesting(CompletionHandler<void(std::optional<uint64_t>)>&&);
 
-    void fetchLocalStorage(CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>)>&&);
+    void fetchLocalStorage(CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&)>&&);
+    void restoreLocalStorage(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&, CompletionHandler<void(bool)>&&);
 
 private:
     enum class ForceReinitialization : bool { No, Yes };

--- a/Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm
+++ b/Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm
@@ -33,49 +33,96 @@
 #import "TestWKWebView.h"
 #import "Utilities.h"
 #import <WebKit/WKScriptMessageHandler.h>
-#import <WebKit/WKWebViewPrivate.h>
 #import <WebKit/WKWebsiteDataRecord.h>
+#import <WebKit/WKWebsiteDataStore.h>
 #import <WebKit/WKWebsiteDataStorePrivate.h>
-#import <WebKit/_WKWebsiteDataStoreConfiguration.h>
 #import <wtf/RetainPtr.h>
 
-static void testFetchLocalStorage(RetainPtr<WKWebsiteDataStore> websiteDataStore)
+static void testRestoreLocalStorage(RetainPtr<WKWebsiteDataStore> websiteDataStore)
 {
+    // Fetch Local Storage.
+
+    // Clear the data store.
+    __block bool done = false;
+    [websiteDataStore removeDataOfTypes:[WKWebsiteDataStore allWebsiteDataTypes] modifiedSince:[NSDate distantPast] completionHandler:^{
+        done = true;
+    }];
+    TestWebKitAPI::Util::run(&done);
+    done = false;
+
+    // Create and load the WebView.
     RetainPtr viewConfiguration = adoptNS([WKWebViewConfiguration new]);
     [viewConfiguration setWebsiteDataStore:websiteDataStore.get()];
     RetainPtr webView = adoptNS([[WKWebView alloc] initWithFrame:CGRectMake(0, 0, 100, 100) configuration:viewConfiguration.get()]);
-
-    auto navigationDelegate = adoptNS([[TestNavigationDelegate alloc] init]);
+    RetainPtr navigationDelegate = adoptNS([[TestNavigationDelegate alloc] init]);
     [webView setNavigationDelegate:navigationDelegate.get()];
+
     [webView loadHTMLString:@"<body>webkit.org</body>" baseURL:[NSURL URLWithString:@"https://webkit.org"]];
     [navigationDelegate waitForDidFinishNavigation];
 
-    __block bool doneEvaluatingJavaScript = false;
+    // Put data into local storage.
     [webView evaluateJavaScript:@"localStorage.setItem('key', 'value');" completionHandler:^(id value, NSError *error) {
         EXPECT_NULL(error);
-        doneEvaluatingJavaScript = true;
+        done = true;
     }];
-    TestWebKitAPI::Util::run(&doneEvaluatingJavaScript);
+    TestWebKitAPI::Util::run(&done);
+    done = false;
 
+    // Fetch the local storage data.
     RetainPtr set = [NSSet setWithObject:WKWebsiteDataTypeLocalStorage];
-
-    __block bool doneFetching = false;
+    __block RetainPtr<NSData> localStorageData;
     [[[webView configuration] websiteDataStore] _fetchDataOfTypes:set.get() completionHandler:^(NSData *data) {
         EXPECT_NOT_NULL(data);
-        EXPECT_GT([data length], 0u);
-        doneFetching = true;
+        localStorageData = data;
+        done = true;
     }];
-    TestWebKitAPI::Util::run(&doneFetching);
+    TestWebKitAPI::Util::run(&done);
+    done = false;
+
+    // Restore Local Storage.
+
+    // Clear the data store.
+    [websiteDataStore removeDataOfTypes:[WKWebsiteDataStore allWebsiteDataTypes] modifiedSince:[NSDate distantPast] completionHandler:^{
+        done = true;
+    }];
+    TestWebKitAPI::Util::run(&done);
+    done = false;
+
+    // Create a new WebView.
+    RetainPtr newWebView = adoptNS([[WKWebView alloc] initWithFrame:CGRectMake(0, 0, 100, 100) configuration:viewConfiguration.get()]);
+    [newWebView setNavigationDelegate:navigationDelegate.get()];
+
+    // Restore the local storage data.
+    [[[newWebView configuration] websiteDataStore] _restoreData:localStorageData.get() completionHandler:^(BOOL succeeded) {
+        EXPECT_TRUE(succeeded);
+        done = true;
+    }];
+    TestWebKitAPI::Util::run(&done);
+    done = false;
+
+    // Load the same page as before.
+    [newWebView loadHTMLString:@"<body>webkit.org</body>" baseURL:[NSURL URLWithString:@"https://webkit.org"]];
+    [navigationDelegate waitForDidFinishNavigation];
+
+    // Check that the local storage data was correctly restored.
+    [newWebView evaluateJavaScript:@"localStorage.getItem('key');" completionHandler:^(id value, NSError *error) {
+        EXPECT_NULL(error);
+        EXPECT_TRUE([value isKindOfClass:[NSString class]]);
+        EXPECT_WK_STREQ(@"value", value);
+        done = true;
+    }];
+    TestWebKitAPI::Util::run(&done);
+    done = false;
 }
 
-TEST(WebKit, FetchLocalStorageFromPersistentDataStore)
+TEST(WebKit, RestoreLocalStorageFromPersistentDataStore)
 {
-    testFetchLocalStorage(adoptNS([WKWebsiteDataStore defaultDataStore]));
+    testRestoreLocalStorage(adoptNS([WKWebsiteDataStore defaultDataStore]));
 }
 
-TEST(WebKit, FetchLocalStorageFromEphemeralDataStore)
+TEST(WebKit, RestoreLocalStorageFromEphemeralDataStore)
 {
-    testFetchLocalStorage(adoptNS([WKWebsiteDataStore nonPersistentDataStore]));
+    testRestoreLocalStorage(adoptNS([WKWebsiteDataStore nonPersistentDataStore]));
 }
 
 @interface RestoreLocalStorageMessageHandler : NSObject <WKScriptMessageHandler>
@@ -106,21 +153,32 @@ function postMessage(message)
 {
     parent.postMessage(message, '*');
 }
-async function setItem()
+function item()
 {
-    try {
+    result = localStorage.getItem('key');
+    if (!result) {
         localStorage.setItem('key', 'value');
         postMessage('Item is set');
-    } catch(err) {
-        postMessage('error: ' + err.name + ' - ' + err.message);
-    }
+    } else
+        postMessage(result);
 }
-setItem();
+item();
 </script>
 )TESTRESOURCE"_s;
 
-static void testFetchLocalStorageThirdPartyIFrame(RetainPtr<WKWebsiteDataStore> websiteDataStore)
+static void testRestoreLocalStorageThirdPartyIFrame(RetainPtr<WKWebsiteDataStore> websiteDataStore)
 {
+    // Fetch Local Storage.
+
+    // Clear the data store.
+    __block bool done = false;
+    [websiteDataStore removeDataOfTypes:[WKWebsiteDataStore allWebsiteDataTypes] modifiedSince:[NSDate distantPast] completionHandler:^{
+        done = true;
+    }];
+    TestWebKitAPI::Util::run(&done);
+    done = false;
+
+    // Create and load the WebView.
     RetainPtr viewConfiguration = adoptNS([[WKWebViewConfiguration alloc] init]);
     RetainPtr handler = adoptNS([[RestoreLocalStorageMessageHandler alloc] init]);
     [[viewConfiguration userContentController] addScriptMessageHandler:handler.get() name:@"testHandler"];
@@ -141,28 +199,57 @@ static void testFetchLocalStorageThirdPartyIFrame(RetainPtr<WKWebsiteDataStore> 
         { "/"_s, { frameBytes } },
     }, TestWebKitAPI::HTTPServer::Protocol::Https, nullptr, nullptr, 9091);
 
+    // The third party iframe will put the data into local storage.
     [webView loadHTMLString:mainFrameString baseURL:[NSURL URLWithString:@"https://webkit.org"]];
     TestWebKitAPI::Util::run(&receivedScriptMessage);
     receivedScriptMessage = false;
     EXPECT_WK_STREQ(@"Item is set", [lastScriptMessage body]);
 
+    // Fetch the local storage data.
     RetainPtr set = [NSSet setWithObject:WKWebsiteDataTypeLocalStorage];
-
-    __block bool doneFetching = false;
+    __block RetainPtr<NSData> localStorageData;
     [[[webView configuration] websiteDataStore] _fetchDataOfTypes:set.get() completionHandler:^(NSData *data) {
         EXPECT_NOT_NULL(data);
-        EXPECT_GT([data length], 0u);
-        doneFetching = true;
+        localStorageData = data;
+        done = true;
     }];
-    TestWebKitAPI::Util::run(&doneFetching);
+    TestWebKitAPI::Util::run(&done);
+    done = false;
+
+    // Restore Local Storage.
+
+    // Clear the data store.
+    [websiteDataStore removeDataOfTypes:[WKWebsiteDataStore allWebsiteDataTypes] modifiedSince:[NSDate distantPast] completionHandler:^{
+        done = true;
+    }];
+    TestWebKitAPI::Util::run(&done);
+    done = false;
+
+    // Create a new WebView.
+    RetainPtr newWebView = adoptNS([[WKWebView alloc] initWithFrame:CGRectMake(0, 0, 100, 100) configuration:viewConfiguration.get()]);
+    [newWebView setNavigationDelegate:navigationDelegate.get()];
+
+    // Restore the local storage data.
+    [[[newWebView configuration] websiteDataStore] _restoreData:localStorageData.get() completionHandler:^(BOOL succeeded) {
+        EXPECT_TRUE(succeeded);
+        done = true;
+    }];
+    TestWebKitAPI::Util::run(&done);
+    done = false;
+
+    // The third party iframe will check that the local storage data was correctly restored.
+    [newWebView loadHTMLString:mainFrameString baseURL:[NSURL URLWithString:@"https://webkit.org"]];
+    TestWebKitAPI::Util::run(&receivedScriptMessage);
+    receivedScriptMessage = false;
+    EXPECT_WK_STREQ(@"value", [lastScriptMessage body]);
 }
 
-TEST(WebKit, FetchLocalStorageFromPersistentDataStoreThirdPartyIFrame)
+TEST(WebKit, RestoreLocalStorageFromPersistentDataStoreThirdPartyIFrame)
 {
-    testFetchLocalStorageThirdPartyIFrame(adoptNS([WKWebsiteDataStore defaultDataStore]));
+    testRestoreLocalStorageThirdPartyIFrame(adoptNS([WKWebsiteDataStore defaultDataStore]));
 }
 
-TEST(WebKit, FetchLocalStorageFromEphemeralDataStoreThirdPartyIFrame)
+TEST(WebKit, RestoreLocalStorageFromEphemeralDataStoreThirdPartyIFrame)
 {
-    testFetchLocalStorageThirdPartyIFrame(adoptNS([WKWebsiteDataStore nonPersistentDataStore]));
+    testRestoreLocalStorageThirdPartyIFrame(adoptNS([WKWebsiteDataStore nonPersistentDataStore]));
 }

